code/campaignclassic/src/main/java/com/adobe/marketing/mobile/CampaignPushTrackerActivity.java

```diff
@@ -0,0 +1,186 @@
+                            + " interaction.");
+            finish();
+            return;
+        }
+        final String action = intent.getAction();
+        if (StringUtils.isNullOrEmpty(action)) {
+            Log.warning(
+                    CampaignPushConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Intent action is null or empty. Ignoring to track or take action on push"
+                            + " notification interaction.");
+            finish();
+            return;
+        }
+
+        switch (action) {
+            case CampaignPushConstants.NotificationAction.OPENED:
+                handlePushOpen(intent);
+                break;
+            case CampaignPushConstants.NotificationAction.BUTTON_CLICKED:
+                handlePushButtonClicked(intent);
+                break;
+            default:
+                break;
+        }
+        finish();
+    }
+
+    /**
+     * Handles the push notification open action.
+     *
+     * @param intent the intent received from the push notification interaction
+     */
+    private void handlePushOpen(final Intent intent) {
+        CampaignClassic.trackNotificationReceive(getTrackInfo(intent));
+        executePushAction(intent);
+    }
+
+    /**
+     * Handles clicks on push notification custom buttons.
+     *
+     * @param intent the intent received from interacting with buttons on push notification
+     */
+    private void handlePushButtonClicked(final Intent intent) {
+        CampaignClassic.trackNotificationClick(getTrackInfo(intent));
+
+        // Dismiss the notification once interacted
+        final String messageId =
+                intent.getStringExtra(CampaignPushConstants.Tracking.Keys.MESSAGE_ID);
+        if (!StringUtils.isNullOrEmpty(messageId)) {
+            NotificationManager notificationManager =
+                    (NotificationManager)
+                            getApplicationContext().getSystemService(Context.NOTIFICATION_SERVICE);
+            notificationManager.cancel(messageId.hashCode());
+        } else {
+            Log.warning(
+                    CampaignPushConstants.LOG_TAG,
+                    SELF_TAG,
+                    "Message ID is null or empty. Unable to dismiss the notification.");
+        }
+
+        executePushAction(intent);
+    }
+
+    /**
+     * Retrieves the Campaign Classic push notification tracking information from the received
+     * notification's {@link Intent}
+     *
+     * @param intent the intent received from the push notification
+     * @return {@link Map<String, String>} containing the notification's tracking information
+     */
+    private Map<String, String> getTrackInfo(final Intent intent) {
+        final Map<String, String> trackInfo = new HashMap<>();
+        final Bundle extras = intent.getExtras();
+        if (extras != null) {
+            for (final String key : extras.keySet()) {
+                final Object value = extras.get(key);
```

Comment:
Can you change this to iterate over the entry set of the map rather than the keys. That way you have both the key and the value at once.

Reply:
the `Bundle` extras only have a `keySet` method.

Reply:
Oh. My bad. I did not realize that this was an extra bundle. Thank you for clarifying!